gdbplus/gdb-9.2/gdb/gdb_dep.c

```diff
@@ -2809,3 +2809,114 @@ ca_eval_address(const char* expr)
 {
 	return parse_and_eval_address(expr);
 }
+
+// return true if an unamed union has a data member of given name
+static bool
+union_has_field_name(struct type *type, const char *field_name)
+{
+	type = check_typedef (type);
+	if (TYPE_CODE(type) != TYPE_CODE_UNION)
+		return false;
+	for (int i = 0; i < TYPE_NFIELDS(type); i++) {
+		const char *name = TYPE_FIELD_NAME(type, i);
+		if (name && *name != '\0') {
+			if (strcmp (field_name, name) == 0)
+				return true;
+		} else if (union_has_field_name(TYPE_FIELD_TYPE(type, i), field_name)) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static int
+type_field_name2no(struct type *type, const char *field_name)
+{
+	if (type == NULL)
+		return -1;
+
+	type = check_typedef (type);
+
+	for (int n = 0; n < TYPE_NFIELDS(type); n++) {
+		const char* name = TYPE_FIELD_NAME(type, n);
+		if (name && *name != '\0') {
+			if (strcmp (field_name, name) == 0)
+				return n;
+		} else if (union_has_field_name(TYPE_FIELD_TYPE(type, n), field_name)) {
+			return n;
+		}
+	}
+	return -1;
+}
+
+bool
+ca_get_field_value(struct value *val, const char *fieldname,
+				size_t *data, bool optional)
+{
+	int fieldno;
+	struct value *field_val;
+
+	*data = ULONG_MAX;
+	fieldno = type_field_name2no(value_type(val), fieldname);
+	if (fieldno < 0) {
+		if (optional) {
+			return true;
+		} else {
+			CA_PRINT("Failed to find member \"%s\"\n", fieldname);
+			return false;
+		}
+	}
+	field_val = value_field(val, fieldno);
+	*data = value_as_long(field_val);
+	return true;
+}
+
+bool
+ca_memcpy_field_value(struct value *val, const char *fieldname, char *buf,
+    size_t bufsz)
+{
+	int fieldno;
+	struct value *field_val;
+	size_t fieldsz;
+
+	fieldno = type_field_name2no(value_type(val), fieldname);
+	if (fieldno < 0) {
+		return false;
+	}
+	field_val = value_field(val, fieldno);
+	fieldsz = TYPE_LENGTH(value_type(field_val));
+	if (bufsz < fieldsz) {
+		CA_PRINT("Internal error: buffer of member \"%s\" is too small\n",
+		    fieldname);
+		return false;
+	}
+	if (!read_memory_wrapper(NULL, value_address(field_val), buf, fieldsz)) {
+		CA_PRINT("Failed to read member \"%s\"\n", fieldname);
+		return false;
+	}
+	return true;
+}
+
+struct value *
+ca_get_field_gdb_value(struct value *val, const char *field_name)
```

Comment:
who is the owner of the returned value? who is responsible to release the resource of the pointer pointed to?

Reply:
I think, and I hope, it is the gdb internal, meaning its symbol/type/value object management engine. I am not 100% sure but I don't see that caller needs to release them. Need more research.